### PR TITLE
Add user auth backend

### LINKED ISSUE #11: Add user auth + backend

### PR DESCRIPTION
**Pull request by nicolehao34**
## Summary

This PR builds out the full LeetCode-style problem-solving experience end-to-end, from clickable problem cards to a real evaluation backend, plus a JWT-based user auth system with server-side persistence.

---

### Problem Solving Page (`/problems/:id`)

- Clicking a problem title navigates to a split-pane page: **Monaco editor** on the right, **4-tab panel** (Description, Editorial, Solutions, Submissions) on the left
- Language selector (Python / JS / TS) with starter code per language
- **Run** button hits `/api/run` → public tests only, fast feedback with per-test pass/fail rows (expandable error tracebacks)
- **Submit** button hits `/api/submit` → private graded tests, shows score breakdown by category
- Graceful fallback: problems without a backend evaluator silently show mock output instead of an error

### Evaluation Backend (`server/`)

- New `evaluator.py`: writes user code to an isolated temp dir, runs `pytest --json-report`, parses per-test results with point scoring from `problem.json`
- Two modes: `run` (public tests) and `submit` (private graded eval script), both subprocess-isolated with configurable timeouts
- First problem wired up: **trans-1 Scaled Dot-Product Attention** — 13 public tests + 15 private graded tests (100 pts across Numerical Accuracy, Masking, Dropout, Edge Cases)

### New Problems & Mind Map Nodes

- **10 new Transformers problems** (trans-6–trans-15): Causal Mask, Numerically Stable Softmax, MQA, GQA, Attention Residual Dropout, FlashAttention, Decoder-Only Block, Encoder-Decoder Block, RMSNorm, Activation Checkpointing
- **5 new VLM problems** (vlm-6–vlm-10): ViT Patch Embedding, Vision Projector, Image Token Inserter, SigLIP Similarity Loss, LLaVA Fusion
- **5 new mind map nodes**: Patch Embedding, SigLIP, LLaVA, Vision Projector, Image Token Inserter (wired to unlock on problem solve)

### User Auth System (JWT + SQLite)

**Backend**
- `database.py` — SQLite via SQLModel, WAL mode for concurrent reads
- `models.py` — 6 tables: `User`, `Submission`, `SolvedProblem`, `ProblemList`, `ProblemListItem`, `RoadmapProgress`
- `auth_utils.py` — bcrypt password hashing, 15-min access JWT + 7-day httpOnly refresh cookie, `get_current_user` / `get_optional_user` FastAPI dependencies
- `routers/auth.py` — `POST /api/auth/register`, `/login`, `/refresh`, `/logout`; `GET /api/auth/me`
- `routers/users.py` — `GET/POST/DELETE /api/users/me/solved/{id}`, `GET /api/users/me/submissions`, full CRUD for `/api/users/me/lists`
- `/api/submit` now persists `Submission` rows for authenticated users; anonymous users still get results

**Frontend**
- `AuthContext.tsx` — `useAuth` hook with `login`, `register`, `logout`, `authFetch` (auto-retries on 401 using refresh endpoint); one-time localStorage migration on first register
- `LoginPage.tsx` / `RegisterPage.tsx` — clean forms, redirect back to original URL on success
- `ProtectedRoute.tsx` — spinner during token validation, redirects to `/login` with `state.from` if unauthenticated
- `App.tsx` — `AuthProvider` wraps app, nav bar shows avatar + username + logout when signed in, `/problems` and `/problems/:id` are protected
- `ProblemsPage.tsx` — solved set and custom lists fetched from server on mount, all mutations hit the API with optimistic UI updates
- `ProblemSolvingPage.tsx` — auth token passed on all API calls, submission history loaded from server

## Test plan

- [ ] `pip install -r requirements.txt && python main.py` — server starts, `db.sqlite` created
- [ ] `POST /api/auth/register` returns access token and sets refresh cookie
- [ ] `GET /api/auth/me` with Bearer token returns user object
- [ ] `POST /api/auth/refresh` (with cookie) issues a new access token
- [ ] Navigating to `/problems` without a session redirects to `/login`
- [ ] After login, `/problems` loads server-fetched solved set
- [ ] Marking a problem solved persists across browser sessions (confirm via Network tab)
- [ ] Running `trans-1` shows per-test pass/fail rows
- [ ] Submitting `trans-1` shows 100-pt score breakdown and saves a Submission row
- [ ] Logout clears session; `/problems` redirects to `/login`
- [ ] Cold-load with a valid token in localStorage restores session without re-login

### CODE REVIEW

File: client/src/pages/ProblemsPage.tsx [hunk 1]
```diff
@@ -185,63 +185,78 @@ export default function ProblemsPage() {
...
-  const deleteList = useCallback((listId: string) => {
-    setMyLists((prev) => prev.filter((l) => l.id !== listId))
+  const deleteList = useCallback(async (listId: string) => {
+    setMyLists(prev => prev.filter(l => l.id !== listId))
     if (selectedListId === listId) setSelectedListId(null)
-  }, [selectedListId])
+    try {
+      await authFetch(`/api/users/me/lists/${listId}`, { method: 'DELETE' })
+    } catch {}
+  }, [selectedListId, authFetch])
```
[the remark this answers is not in the document]
Comment (nicolehao34) [reply client/src/pages/ProblemsPage.tsx:259]: @copilot apply changes based on [this feedback](https://github.com/nicolehao34/InkSight-Deep-Learning-Roadmap/pull/13#discussion_r3112871123)

File: client/src/pages/ProblemsPage.tsx [hunk 2]
```diff
@@ -185,63 +185,78 @@ export default function ProblemsPage() {
...
+    setMyLists(prev =>
+      prev.map(l =>
+        l.id === listId ? { ...l, problemIds: l.problemIds.filter(id => id !== problemId) } : l
       )
     )
-  }, [])
+    try {
+      await authFetch(`/api/users/me/lists/${listId}/items/${problemId}`, { method: 'DELETE' })
+    } catch {}
+  }, [authFetch])
```
[the remark this answers is not in the document]
Comment (nicolehao34) [reply client/src/pages/ProblemsPage.tsx:251]: @copilot apply changes based on [this feedback](https://github.com/nicolehao34/InkSight-Deep-Learning-Roadmap/pull/13#discussion_r3112871106)
